daemon/cmd/app.go

```diff
@@ -0,0 +1,224 @@
+				HostDevice: defaults.HostDevice,
+				ProcFs:     config.ProcFs,
+			}
+		}),
+
+	fx.Provide(
+		linuxdatapath.NewDatapath,
+	),
+)
+
+var daemonModule = fx.Module(
+	"daemon",
+
+	fx.Provide(
+		NewDaemon,
+
+		// Lift out some objects from the daemon for modules.
+		func(d *Daemon) (*ipcache.IPCache, mtu.Configuration) {
+			return d.ipcache, d.mtuConfig
+		},
+	),
+
+	fx.Invoke(
+		preInit,
+		registerDaemonStart,
+		endpointManagerInit,
+	),
+)
+
+func runApp(cmd *cobra.Command) {
+	// Create a top-level context for the daemon. This is cancelled by the signal handler
+	// in cleanup.go and will trigger stop.
+	ctx, cancel := context.WithCancel(server.ServerCtx)
+
+	initEnv(cmd)
+
+	app := fx.New(
+		fx.StartTimeout(startTimeout),
+		fx.StopTimeout(stopTimeout),
+		fx.WithLogger(newAppLogger),
+
+		fx.Supply(
+			fx.Annotate(ctx, fx.As(new(context.Context))),
+			cancel),
+
+		fx.Supply(option.Config),
+
+		fx.Provide(
+			iptables.NewIptablesManager,
+			newEndpointManager,
+		),
+
+		// The order in which the modules are declared denotes the invoke and
+		// OnStart order.
+		linuxDatapathModule,
+		daemonModule,
+		wireguard.Module,
+		optional(option.Config.EnableIPMasqAgent, ipmasqAgentModule),
+
+		fx.Invoke(writeDotGraph),
+	)
+
+	if app.Err() != nil {
+		log.WithError(app.Err()).Fatal("Failed to initialize daemon")
+	}
+
+	if option.Config.DotGraphOutputFile != "" {
+		log.WithField("file", option.Config.DotGraphOutputFile).Infof("Wrote dot graph and now exiting without starting")
+		os.Exit(0)
+	}
+
+	startCtx, cancel := context.WithTimeout(context.Background(), app.StartTimeout())
+	defer cancel()
+
+	if err := app.Start(startCtx); err != nil {
+		if ctx.Err() != nil {
+			// Context was cancelled during startup, e.g. due to interrupt. Exit
+			// normally in this case.
+			os.Exit(0)
+		}
+		os.Exit(1)
+	}
+
+	// Wait until the daemon context is cancelled. This is done via the
+	// cleaner on receipt of an interrupt.
+	<-ctx.Done()
+
+	stopCtx, cancel := context.WithTimeout(context.Background(), app.StopTimeout())
+	defer cancel()
+
+	if err := app.Stop(stopCtx); err != nil {
+		log.WithError(err).Fatal("Failed to stop daemon")
+	}
+}
+
+func optional(flag bool, opts ...fx.Option) fx.Option {
+	if flag {
+		return fx.Options(opts...)
+	}
+	return fx.Invoke()
+}
+
+func preInit() error {
+	log.Info("Initializing daemon")
+	option.Config.RunMonitorAgent = true
+
+	if err := enableIPForwarding(); err != nil {
+		return fmt.Errorf("error when enabling sysctl parameters: %w", err)
+	}
+	if k8s.IsEnabled() {
+		bootstrapStats.k8sInit.Start()
+		if err := k8s.Init(option.Config); err != nil {
+			return fmt.Errorf("unable to initialize Kubernetes subsystem: %w", err)
+		}
+		bootstrapStats.k8sInit.End(true)
+	}
+	return nil
+}
+
+func registerDaemonStart(lc fx.Lifecycle, d *Daemon, restoredEndpoints *endpointRestoreState) {
+	lc.Append(fx.Hook{
+		OnStart: func(context.Context) error {
+			go runDaemon(d, restoredEndpoints)
```

Comment:
I think we probably need to take a closer look at the parallelism here, this looks to me like it was initially added because the `runDaemon()` function today will block until the agent receives a signal, whereas `fx` probably doesn't expect this from an `OnStart()` function. It seems like there was another solution made to this with the error channels, but I'm not exactly sure just yet.